ibis/expr/operations/udf.py

```diff
@@ -51,21 +57,130 @@ def _wrap(
     return functools.update_wrapper(wrapper(input_type, fn, *args, **kwargs), fn)
 
 
-S = TypeVar("S", bound=ScalarUDF)
+S = TypeVar("S", bound=ops.Value)
+B = TypeVar("B", bound=ops.Value)
+
+
+class _UDF(abc.ABC):
+    __slots__ = ()
+
+    @property
+    @abc.abstractmethod
+    def base(self) -> type[B]:
+        """Base class of the UDF."""
+
+    @util.experimental
+    @classmethod
+    def builtin(
+        cls,
+        fn: Callable | None = None,
+        *args: Any,
+        name: str | None = None,
+        schema: str | None = None,
+        **kwargs: Any,
+    ) -> Callable:
+        """Construct a scalar user-defined function that is built-in to the backend.
```

Comment:
This docstring will show up for both `ibis.udf.scalar.builtin` and `ibis.udf.agg.builtin`, but says "scalar" above (and shows a scalar example below). Perhaps this belongs on the subclasses rather than the base class? The implementation is short enough that repeating the code feels fine to me.

Reply:
Repeated the docstring and made it `agg`-worthy